### PR TITLE
implement patch extraction in ops

### PR DESCRIPTION
**Pull request by anas-rz**
As discussed in #529, this pull request implements [tf.image.extract_patches](https://www.tensorflow.org/api_docs/python/tf/image/extract_patches)  in keras ops

### CODE REVIEW

File: keras_core/ops/image.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 from keras_core.backend import KerasTensor
 from keras_core.backend import any_symbolic_tensors
 from keras_core.ops.operation import Operation
-
+from keras_core import ops
```
Comment (fchollet) [inline keras_core/ops/image.py:6]: This sounds like a circular import. Import `backend` instead.

File: keras_core/ops/image.py [hunk 2]
```diff
@@ -244,3 +244,126 @@ def affine_transform(
...
+
+
+@keras_core_export("keras_core.ops.image.extract_patches")
+def extract_patches(
+    image, sizes, strides, rates, padding, data_format="channels_last", name=None,
+    
+):
+    """Applies the given transform(s) to the image(s).
+
+    Args:
```
Comment (fchollet) [inline keras_core/ops/image.py:292]: Fix the args section.

File: keras_core/ops/image.py [hunk 3]
```diff
@@ -244,3 +244,126 @@ def affine_transform(
...
+            while `"channels_first"` corresponds to inputs with shape
+            `(batch, channels, height, weight)`. It defaults to the
+            `image_data_format` value found in your Keras config file at
+            `~/.keras/keras.json`. If you never set it, then it will be
+            `"channels_last"`.
+
+    Returns:
+        Applied affine transform image or batch of images.
+
+    Examples:
```
Comment (fchollet) [inline keras_core/ops/image.py:324]: Fix the examples section.

File: keras_core/ops/image.py [hunk 4]
```diff
@@ -244,3 +244,126 @@ def affine_transform(
...
+                "Invalid transform rank: expected rank 1 (single transform) "
+                "or rank 2 (batch of transforms). Received input with shape: "
+                f"transform.shape={transform.shape}"
+            )
+        return KerasTensor(image.shape, dtype=image.dtype)
+
+
+@keras_core_export("keras_core.ops.image.extract_patches")
+def extract_patches(
+    image, sizes, strides, rates, padding, data_format="channels_last", name=None,
```
Comment (fchollet) [inline keras_core/ops/image.py:287]: sizes -> size

Comment (fchollet) [reply keras_core/ops/image.py:287]: No `name` argument

File: keras_core/ops/image.py [hunk 5]
```diff
@@ -244,3 +244,126 @@ def affine_transform(
...
+                "or rank 4 (batch of images). Received input with shape: "
+                f"image.shape={image.shape}"
+            )
+        if len(transform.shape) not in (1, 2):
+            raise ValueError(
+                "Invalid transform rank: expected rank 1 (single transform) "
+                "or rank 2 (batch of transforms). Received input with shape: "
+                f"transform.shape={transform.shape}"
+            )
+        return KerasTensor(image.shape, dtype=image.dtype)
```
Comment (fchollet) [inline keras_core/ops/image.py:282]: It should return the output shape

File: keras_core/ops/image.py [hunk 6]
```diff
@@ -244,3 +244,126 @@ def affine_transform(
...
+    if data_format == 'channels_last':
+        channels_in = image.shape[-1]
+        out_dim = image.shape[1] // patch_h * image.shape[2] // patch_w
+    elif data_format == 'channels_first':
+        # if batched
+        channels_in = image.shape[1]
+        out_dim = (image.shape[2] // patch_h) * (image.shape[3] // patch_w)
+    kernel = ops.eye(patch_h * patch_w * channels_in, out_dim)
+    kernel = ops.reshape(kernel, (patch_h, patch_w, channels_in, out_dim))
+    return ops.nn.conv(image, kernel, 14, 'VALID')
```
Comment (fchollet) [inline keras_core/ops/image.py:369]: - Padding should be lowercase
- Prefer using keyword arguments for better readbility

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

Please add unit tests as well to check shape inference and output correctness.

File: keras_core/ops/image.py [hunk 7]
```diff
@@ -244,3 +244,216 @@ def affine_transform(
...
+        return _extract_patches(
+            image,
+            self.size,
+            self.strides,
+            self.rates,
+            self.padding,
+            data_format=self.data_format,
+        )
+
+    def compute_output_spec(self, image):
```
Comment (fchollet) [inline keras_core/ops/image.py:275]: You should be able to reuse the same utilities that are used to compute the output spec of the `Conv` operation.

**Comment by anas-rz** [comment]
Updating doc string

File: keras_core/ops/image.py [hunk 8]
```diff
@@ -244,3 +245,177 @@ def affine_transform(
...
+        self.padding = padding
+        self.data_format = data_format
+
+    def call(self, image):
+        return _extract_patches(
+            image,
+            self.size,
+            self.strides,
+            self.rates,
+            self.padding,
```
Comment (fchollet) [inline keras_core/ops/image.py:272]: Prefer using keyword arguments for better readability

File: keras_core/ops/image.py [hunk 9]
```diff
@@ -244,3 +245,177 @@ def affine_transform(
...
         data_format=data_format,
     )
+
+
+class ExtractPatches(Operation):
+    def __init__(
+        self,
+        size,
+        strides=None,
+        rates=1,
```
Comment (fchollet) [inline keras_core/ops/image.py:255]: Is this the same as the `dilation_rate` arg in conv? If so, it should be renamed `dilation_rate`.

File: keras_core/ops/image.py [hunk 10]
```diff
@@ -244,3 +245,177 @@ def affine_transform(
...
+    Examples:
+
+    >>> x = np.random.random((2, 64, 80, 3)) # batch of 2 RGB images
+    >>> transform = np.array(
+    ...     [
+    ...         [1.5, 0, -20, 0, 1.5, -16, 0, 0],  # zoom
+    ...         [1, 0, -20, 0, 1, -16, 0, 0],  # translation
+    ...     ]
+    ... )
+    >>> y = keras_core.ops.image.affine_transform(x, transform)
```
Comment (fchollet) [inline keras_core/ops/image.py]: Same for the examples section -- this is from `affine_transform`

File: keras_core/ops/image.py [hunk 11]
```diff
@@ -244,3 +245,177 @@ def affine_transform(
...
+    padding="valid",
+    data_format="channels_last",
+):
+    if type(size) == int:
+        patch_h = patch_w = size
+    elif len(size) == 2:
+        patch_h, patch_w = size[0], size[1]
+    else:
+        raise TypeError(
+            "Received invalid patch size expected "
```
Comment (fchollet) [inline keras_core/ops/image.py:371]: "invalid `size` argument"

File: keras_core/ops/image.py [hunk 12]
```diff
@@ -244,3 +245,177 @@ def affine_transform(
...
+            f"int or tuple of length 2, received {size}"
```
Comment (fchollet) [inline keras_core/ops/image.py:372]: ". Received: size={size}"

**Review by fchollet — COMMENTED** [review]
Looks great! Thanks!

**Comment by anas-rz** [comment]
Fixed. CC  @fchollet

**Review by fchollet — APPROVED** [review]
Thanks for the contribution!

**Comment by anas-rz** [comment]
Thank you for your guidance. Can I add [tf.extract_volume_patches](https://www.tensorflow.org/api_docs/python/tf/extract_volume_patches) to keras_core.ops?
